pkg/util/query/priority.go

```diff
@@ -0,0 +1,87 @@
+package query
+
+import (
+	"math"
+	"net/url"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+func GetPriority(requestParams url.Values, now time.Time, queryPriority *validation.QueryPriority, queryPriorityChanged bool) int64 {
+	queryParam := requestParams.Get("query")
+	timeParam := requestParams.Get("time")
+	startParam := requestParams.Get("start")
+	endParam := requestParams.Get("end")
+
+	if queryParam == "" || !queryPriority.Enabled {
+		return queryPriority.DefaultPriority
+	}
+
+	for i, priority := range queryPriority.Priorities {
+		for j, attribute := range priority.QueryAttributes {
+			// If query priority config changed, re-populate the compiled regex
+			if queryPriorityChanged {
+				compiledRegex, err := regexp.Compile(attribute.Regex)
+				if err != nil {
+					continue
+				}
+
+				attribute.CompiledRegex = compiledRegex
+				queryPriority.Priorities[i].QueryAttributes[j] = attribute
+			}
+
+			if attribute.CompiledRegex != nil && !attribute.CompiledRegex.MatchString(queryParam) {
+				continue
+			}
+
+			startTimeThreshold := now.Add(-1 * attribute.StartTime.Abs()).Truncate(time.Second).UTC()
+			endTimeThreshold := now.Add(-1 * attribute.EndTime.Abs()).Add(1 * time.Second).Truncate(time.Second).UTC()
```

Comment:
If only start time is set but end time is not set, then end time threshold will be now.
Then if users set end time to a future timestamp, it won't have a match. Shall we have a default start/end time rather than using now?

Reply:
Are you suggesting to cap the end time to be now, even though user queries for future timestamp?

Reply:
Ah the opossite. I see even if there is no end time set in query priority, we cap it to now? Is it better to not cap when start/end time not set? Maybe I misunderstood the code

Reply:
gotcha, yup it makes sense to not check at all if they're not specified

Reply:
Addressed in https://github.com/cortexproject/cortex/pull/5605/commits/78816d94ad6c1bca9e310d77e6a7405a81ec2565